Bluetooth: Mesh: Decouple blob_io_flash from flash

ludvigsj opened the pull request:
This fully decouples blob_io_flash from flash configuration details, by removing remaining direct references to CONFIG_FLASH_* options and hard-coded write block size, instead pulling this from the actual flash device used.

It also requires the use of FLASH_PAGE_LAYOUT APIs when enabling the module for use with flash with explicit erase. This was already a silent requirement, since the existing code does not work for such use cases except for the case when a BLOB has a size which is a multiple of the page size (since, without it, trying to erase flash space for the final block, which is of variable size, will fail).

Additionally, the erasure logic has been changed slightly to only erase a page when starting the first block on that page. This fixes a bug where, if the block size is smaller than the page size, the entire page would be erased upon start of each block when writing, meaning only the final block on each page would be retained. This works because blocks are always received in order.

(Note: we might need to add some code to ensure that the area and offset passed to the init function lies on a page boundary. I suggest to defer that to a separate PR, as it is an existing problem already and not introduced by my changes)

The block write routine has been updated to be more generic and fix a few bugs:

  * The write buffer is now sized based on a KConfig option instead of basing it off the device tree - this makes the module usable with any flash device, not just the internal memory. The configured value is checked during initialization, to ensure that the configured size will fit the write block size required by the device passed to the init function.
  * The erase value used to fill the buffer when using a device with explicit erase is now pulled from the flash parameters instead of being hard-coded to `0xff`.
  * An additional write block sized piece of buffer is allocated - the previous buffer sizing with rounding up only worked if `BLOB_RX_CHUNK_SIZE % WRITE_BLOCK_SIZE` was 0 or 1 (which coincidentally worked in all testing because the chunk size defaults to 161, and for internal flash w/write block size of 4, `161 % 4 == 1`).
  * The choice of whether to just write the chunk as-is or to write-block align it is now based on the erase cap pulled from the flash parameters, instead of checking the type of memory in the SOC. This means the module dos _not_ currently support the case where memory has to be written write-block aligned, but the memory does _not_ use explicit erase. It uses a trick of filling the write buffer with the erase value to avoid overwriting existing chunks, which will not work in this case. This trick is required to support random ordered chunks while still writing write-block aligned.
  * The code has been cleaned up a bit in general.

ludvigsj commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -1058,12 +1058,24 @@ config BT_MESH_BLOB_IO_FLASH_WITH_ERASE
+
+config BT_MESH_BLOB_IO_FLASH_WRITE_BLOCK_SIZE_MAX
+	int "Maximum supported write block size"
+	default 16
```
Right, I see. I set it to 16 because RRAM on nRF54L uses 16, but that of course doesn't need a buffer at all since it is written directly, my mistake.

PavelVPV replied:
Then default value could be based on platform, but either way fine by my to keep as is considering the reply above.

PavelVPV replied:
> This option does not affect the RRAM case at all, because that does not use a buffer. The only thing this option does is control the buffer size. It does not control the writing, it does not control alignment.

Oh, yes, you are right. In case of RRAM this will be executed:https://github.com/zephyrproject-rtos/zephyr/blob/b77724229f8de62db5ac454c6822289b22312150/subsys/bluetooth/mesh/blob_io_flash.c#L150-L154

ludvigsj replied:
Oh, you are right. Then I still think the current behavior (which, bear in mind, is unchanged compared to the behavior on the current main branch) is the best we can do without a larger rethink/rewrite. As I mentioned, just writing everything unbuffered to RRAM will (assuming the chunk size is always much larger than the write block size) cause it to wear out twice as fast in the worst case (the write blocks on chunk boundaries will be written twice). Since BLOB reception is typically not happening with a very high frequency, I think this should be fine for now.

ludvigsj commented on subsys/bluetooth/mesh/blob_io_flash.c:
```diff
@@ -56,42 +64,42 @@ static void io_close(const struct bt_mesh_blob_io *io,
 }
 
+/* Erasure code not needed if no flash in the system requires explicit erase */
+#ifdef CONFIG_BT_MESH_BLOB_IO_FLASH_WITH_ERASE
```
There is no point in using `IS_ENABLED`, checking of the flash params from the driver is enough in that case, so the only reason to keep using these KConfigs would be conditional compilation. If we don't want to do that, they can be completely removed to simplify, optionally only used when assigning the handler.

ludvigsj replied:
> prefer to not change code

I don't see why these options were introduced in the first place if not to save code size when erasure is not needed.

omkar3141 replied:
I don't think LTO is a mainstream option yet. It is not enabled by default nor we recommend it anywhere.

ludvigsj commented on subsys/bluetooth/mesh/blob_io_flash.c:
```diff
@@ -125,30 +134,49 @@ static int wr_chunk(const struct bt_mesh_blob_io *io,
+	 * Allocate one additional write block for the case where a chunk will need
+	 * an extra write block on both sides to fit.
+	 */
+	uint8_t buf[ROUND_UP(BLOB_RX_CHUNK_SIZE, CONFIG_BT_MESH_BLOB_IO_FLASH_WRITE_BLOCK_SIZE_MAX)
```
I decided to work with the principle of as few changes as possible here. If we want this as an optimization, I think it should be a separate task. The buffer is currently on the stack on _main_.

(Also, using the stack means you can save quite a bit of memory if you have an application with multiple BLOB IO contexts initialized at the same time (to different flash areas) but only use one of them in any given transfer)

ludvigsj replied:
https://github.com/zephyrproject-rtos/zephyr/blob/0748b4e47d1f2888708b65d0a7ed5a2d10851b62/subsys/bluetooth/mesh/blob_io_flash.c#L135 Stack allocated buffer, current main. I did not introduce this.

omkar3141 commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -1058,12 +1058,24 @@ config BT_MESH_BLOB_IO_FLASH_WITH_ERASE
 	bool "BLOB flash support for devices with erase"
```
> In either case, I would consider their removal (if we decide to do so) as a separate PR.

Why not in this one, since it is relavant for the title of the PR?

PavelVPV replied:
> > In either case, I would consider their removal (if we decide to do so) as a separate PR.
> 
> Why not in this one, since it is relavant for the title of the PR?

Yeah, up to the author. I'm fine with deprecating these options.

omkar3141 reviewed (commented):
PR looks fine to me. Are we going to remove the KConfig options then?

ludvigsj committed:
Bluetooth: Mesh: Decouple blob_io_flash erase from flash config

This adds a Kconfig dependency on FLASH_PAGE_LAYOUT when using flash
with explicit erase, removing all code dependencies on flash
configuration from the erasure logic. This module already requires
FLASH_PAGE_LAYOUT to be set when using it with flash with explicit
erase, as a silent requirement. The current code does not work without
this option enabled except for the case where a BLOB has a size which is
a multiple of the page size (since, without it, trying to erase flash
space for the final block, which is of variable size, will fail). Also
cleans up the code a bit.

Signed-off-by: Ludvig Jordet <ludvig.jordet@nordicsemi.no>

ludvigsj committed:
Bluetooth: Mesh: make blob_io_flash support smaller blocks

Only erase a page when starting the first block on that page.
This fixes a bug where, if the block size is smaller than the page size,
the entire page would be erased upon start of each block when writing,
meaning only the final block on each page would be retained. This works
because blocks are always received in order.

Signed-off-by: Ludvig Jordet <ludvig.jordet@nordicsemi.no>

ludvigsj committed:
Bluetooth: Mesh: decouple blob_io_flash from internal flash

This fully decouples blob_io_flash from specific flash device details,
by not reading the write block size from a hard-coded device tree node,
instead pulling this from the actual flash device used.

The block write routine has been updated to be more generic and fix a
few bugs:

  * The write buffer is now sized based on a KConfig option instead of
    basing it off the device tree - this makes the module usable with
    any flash device, not just the internal memory. The configured value
    is checked during initialization, to ensure that the configured size
    will fit the write block size required by the device passed to the
    init function.
  * The erase value used to fill the buffer when using a device with
    explicit erase is now pulled from the flash parameters instead of
    being hard-coded to `0xff`.
  * An additional write block sized piece of buffer is allocated - the
    previous buffer sizing with rounding up only worked if
    `BLOB_RX_CHUNK_SIZE % WRITE_BLOCK_SIZE` was 0 or 1 (which
    coincidentally worked in all testing because the chunk size defaults
    to 161, and for internal flash w/write block size of 4,
    `161 % 4 == 1`).
  * The choice of whether to just write the chunk as-is or to
    write-block align it is now based on the erase cap pulled from the
    flash parameters, instead of checking the type of memory in the SOC.
    This means the module dos _not_ currently support the case where
    memory has to be written write-block aligned, but the memory does
    _not_ use explicit erase. It uses a trick of filling the write
    buffer with the erase value to avoid overwriting existing chunks,
    which will not work in this case. This trick is required to support
    random ordered chunks while still writing write-block aligned.
  * The code has been cleaned up a bit in general.

Signed-off-by: Ludvig Jordet <ludvig.jordet@nordicsemi.no>

ludvigsj committed:
Bluetooth: Mesh: Deprecate blob_io_flash erase cap options

The only reason to keep these would be to enable conditional compilation
of some parts of the code. However, we can't see enough benefit in doing
this, particularly since the flash driver itself will conditionally
handle certain paths depending on erase caps, so the options are
deprecated to reduce complexity.

Signed-off-by: Ludvig Jordet <ludvig.jordet@nordicsemi.no>
